test/mocha/13-keystores-update-api.js

```diff
@@ -0,0 +1,295 @@
+      let result;
+      const config = {
+        id: 'foo',
+        controller: 'bar',
+        sequence: 1.1,
+      };
+      try {
+        result = await keystores.update({config});
+      } catch(e) {
+        err = e;
+      }
+      should.not.exist(result);
+      should.exist(err);
+      err.message.should.contain(
+        'Keystore config sequence must be an integer greater than zero.');
+    });
+    it('throws error on zero config.sequence', async () => {
+      let err;
+      let result;
+      const config = {
+        id: 'foo',
+        controller: 'bar',
+        sequence: 0,
+      };
+      try {
+        result = await keystores.update({config});
+      } catch(e) {
+        err = e;
+      }
+      should.not.exist(result);
+      should.exist(err);
+      err.message.should.contain(
+        'Keystore config sequence must be an integer greater than zero.');
+    });
+    it('throws error on string config.sequence', async () => {
+      let err;
+      let result;
+      const config = {
+        id: 'foo',
+        controller: 'bar',
+        sequence: '0',
+      };
+      try {
+        result = await keystores.update({config});
+      } catch(e) {
+        err = e;
+      }
+      should.not.exist(result);
+      should.exist(err);
+      err.message.should.contain('config.sequence (number) is required');
+    });
+    it('throws error on non-string config.id', async () => {
+      let err;
+      let result;
+      const config = {
+        id: 1,
+        controller: 'bar',
+        sequence: '0',
+      };
+      try {
+        result = await keystores.update({config});
+      } catch(e) {
+        err = e;
+      }
+      should.not.exist(result);
+      should.exist(err);
+      err.message.should.contain('config.id (string) is required');
+    });
+    it('throws error on non-string config.controller', async () => {
+      let err;
+      let result;
+      const config = {
+        id: 'foo',
+        controller: 1,
+        sequence: '0',
+      };
+      try {
+        result = await keystores.update({config});
+      } catch(e) {
+        err = e;
+      }
+      should.not.exist(result);
+      should.exist(err);
+      err.message.should.contain('config.controller (string) is required');
+    });
+    it('successfully updates a keystore', async () => {
+      let err;
+      let result;
+      const config = clone(mockConfigAlpha);
+      config.sequence++;
+      // FIXME: should updating controller be allowed?
+      config.controller = 'someOtherController';
+      try {
+        result = await keystores.update({config});
+      } catch(e) {
+        err = e;
+      }
+      assertNoError(err);
+      result.should.be.a('boolean');
+      result.should.be.true;
+    });
+    it('successfully updates a keystore twice', async () => {
+      let err;
+      let result;
+      const config = clone(mockConfigBeta);
+      config.sequence++;
+      // FIXME: should updating controller be allowed?
```

Comment:
@dlongley same fixme in a number of places.  Should `controller` be a protected field?

Reply:
Updating a controller needs to be allowed *somehow* -- it's important for key recovery cases.

Reply:
@dlongley how about an API flag for `allowControllerUpdate` or the like here which defaults to false which would protect the field.  And for those use case that need to overwrite it, they must explicitly set that flag.

Reply:
@mattcollier Ok, we can do that and try out the pattern.